### PR TITLE
invoke cuda kernels with correct stream

### PR DESCRIPTION
**Pull request by krishna-zen**
- Invoke cuda kernels in torchsparse with the right stream argument

NOTE: grepped for all kernel invocations using ">>>" and replaced all. I don't think I missed any, please let me know. (nsys also shows the same)

Without streams:
Screenshot 2025-08-13 at 3 29 56 PM

With streams:
Screenshot 2025-08-13 at 3 29 28 PM

With streams and threading:
Screenshot 2025-08-13 at 3 42 38 PM

After this change, we can see that in radar processor, the common preprocessing still happens on the default stream but the model specific stuff happens on the dedicated stream.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -8,6 +8,24 @@ verify_ssl = true
...
+]
+nvidia-nccl-cu12 = [
+  { index = "pytorch-cu128" }
+]
+nvidia-cuda-cupti-cu12 = [
+  { index = "pytorch-cu128" }
+]
+torchvision = [
+  { index = "pytorch-cu128" }
+]
```
Comment (IlanGuenet) [inline pyproject.toml]: Do you need this to access be able to use the `Aten` library?

File: torchsparseplusplus/backend/others/count_cuda.cu [hunk 2]
```diff
@@ -16,7 +17,8 @@ __global__ void count_kernel(int N, const int *__restrict__ data,
 }
 
 void count_wrapper(int N, const int *data, int *out) {
-  count_kernel<<<ceil((double)N / 512), 512>>>(N, data, out);
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream().stream();
```
Comment (IlanGuenet) [inline torchsparseplusplus/backend/others/count_cuda.cu:20]: How is the current cuda stream set?

**Review by IlanGuenet — APPROVED** [review]
Looks good :+1:

File: pyproject.toml (hunk 1 above)
Comment (traskzen) [reply pyproject.toml]: If it's all the same to you I'd cut out this `uv` stuff and I'll make a PR to move from pdm -> uv more officially.  We need this change but also in the other branch and the README's both need to be updated/tested.

Unless, of course, you need the uv change for something you've done here.

File: torchsparseplusplus/backend/others/count_cuda.cu (hunk 2 above)
Comment (krishna-zen) [reply torchsparseplusplus/backend/others/count_cuda.cu:20]: https://github.com/ZendarInc/RadarProcessor/blob/44e3bd1a7a86e283df68f2dfe1a6f6edab61e523/python/zendar/entrypoints/live_ml_pipeline.py#L261-L263

Comment (krishna-zen) [reply torchsparseplusplus/backend/others/count_cuda.cu:20]: it's kinda similar to the way we handle our streams in our RP cuda code.

File: pyproject.toml (hunk 1 above)
Comment (krishna-zen) [reply pyproject.toml]: @traskzen I only need this to make stuff work. I'll probably wait for #29 and #30 to be merged before this.

Comment (traskzen) [reply pyproject.toml]: @krishna-zen Feel free to LMK what you think of those.  You're as familiar as anyone else with it I would think.  If it looks good I'll merge it in the morning.

File: tests/test.py [hunk 3]
```diff
@@ -47,5 +49,67 @@ def test_to_dense(self):
...
+            acc_adiff = 0.0
+            acc_rdiff = 0.0
+            count = 0
+
+            # hashmap mode by default
+            for kernel_size in kernel_sizes:
+                for stride in strides:
+                    mean_adiff, max_rdiff = test_single_layer_convolution_forward(
+                        kernel_size=kernel_size, stride=stride
+                    )
```
Comment (IlanGuenet) [inline tests/test.py:68]: Nice to have additional unit tests :)
Is it the cuda stream synchronize implicit in `test_single_layer_convolution_forward` ?

Comment (krishna-zen) [reply tests/test.py:68]: yes, I see a `.detach()` So, synchronization should be implicit.

**Comment by krishna-zen** [comment]
@traskzen anyway to skip precommit on these? or something that doesn't involve blowing up the diff unnecessarily?

**Comment by traskzen** [comment]
> @traskzen anyway to skip precommit on these? or something that doesn't involve blowing up the diff unnecessarily?

@krishna-zen I have 2 suggestions:
1) disable the pre-commit check in the `.pre-commit-config.yaml` file in this PR and we can create a ticket to "tidy" the whole repo in a separate commit;
2) just go ahead and let it run on the files you're touching.  You could do it in a separate commit from your changes (before, presumably) and only look at the diffs from your changes.

Option 1 is probably the better way to go but either is fine with me.

**Comment by IlanGuenet** [comment]
I wouldn't recommend `"tidy" the whole repo in a separate commit;` because it will break all the git history, git diff with the original repository.

**Comment by krishna-zen** [comment]
@traskzen I see that you've commented the clang format. Even with that commented, there's still a lot of diff from other formatters which is messing things up.

Should we just not do pre-commit at all?

**Comment by traskzen** [comment]
> @traskzen I see that you've commented the clang format. Even with that commented, there's still a lot of diff from other formatters which is messing things up.
> 
> Should we just not do pre-commit at all?

If you think it is best to just get rid of it all that is fine with me.  From my limited testing it seemed like the style-checks were creating pretty minimal changes (mostly stray whitespace getting removed).  I think the cost vs. benefit calculation is so close it doesn't really matter which way it goes.  

Just remove whatever you want and put it as a part of the PR. 

My measure would be this:  are your changes easily identifiable in the code or not?  If not, disable the linting/style stuff.  If yes, then keep it.

**Comment by zensy** [comment]
@krishna-zen can you show the nsys image (before/after) since you have a tag already? just to confirm

**Comment by krishna-zen** [comment]
Limited precommit checks to minmize diff
